flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreExpireImpl.java

```diff
@@ -0,0 +1,155 @@
+
+    @Override
+    public void expire() {
+        Long latestSnapshotId = pathFactory.latestSnapshotId();
+        if (latestSnapshotId == null) {
+            // no snapshot, nothing to expire
+            return;
+        }
+
+        // binary search for the last snapshot to expire
+        long currentMillis = System.currentTimeMillis();
+        long head = Math.max(Snapshot.FIRST_SNAPSHOT_ID - 1, latestSnapshotId - numRetained);
+        long tail = latestSnapshotId;
+        while (head < tail) {
+            long mid = (head + tail + 1) / 2;
+            Path snapshotPath = pathFactory.toSnapshotPath(mid);
+            try {
+                if (!snapshotPath.getFileSystem().exists(snapshotPath)
+                        || currentMillis - Snapshot.fromPath(snapshotPath).timeMillis()
+                                > millisRetained) {
+                    head = mid;
+                } else {
+                    tail = mid - 1;
+                }
+            } catch (IOException e) {
+                throw new RuntimeException(
+                        "Failed to determine if snapshot #" + mid + " still exists", e);
+            }
+        }
+
+        // determine the exact expire range, we also keep at least 1 snapshot not to expire
+        long lastSnapshotIdToExpire = Math.min(head, latestSnapshotId - 1);
+        long firstSnapshotIdToExpire = Snapshot.FIRST_SNAPSHOT_ID;
+        for (long id = lastSnapshotIdToExpire; id >= Snapshot.FIRST_SNAPSHOT_ID; id--) {
+            Path snapshotPath = pathFactory.toSnapshotPath(id);
+            try {
+                if (!snapshotPath.getFileSystem().exists(snapshotPath)) {
+                    // only latest snapshots are retained, as we cannot find this snapshot, we can
+                    // assume that all snapshots preceding it have been removed
+                    firstSnapshotIdToExpire = id + 1;
+                    break;
+                }
+            } catch (IOException e) {
+                throw new RuntimeException(
+                        "Failed to determine if snapshot #" + id + " still exists", e);
+            }
+        }
+
+        // expire each snapshot
+        for (long id = firstSnapshotIdToExpire; id <= lastSnapshotIdToExpire; id++) {
+            expire(id);
+        }
+    }
+
+    private void expire(long snapshotId) {
+        Snapshot toExpire = Snapshot.fromPath(pathFactory.toSnapshotPath(snapshotId));
+        Snapshot nextSnapshot = Snapshot.fromPath(pathFactory.toSnapshotPath(snapshotId + 1));
+
+        // if sst file in only used in snapshot to expire but not in next snapshot we can delete it
+        // because each sst file will only be added and deleted once
+        Set<Path> sstInUse = new HashSet<>();
+        FileStorePathFactory.SstPathFactoryCache sstPathFactoryCache =
+                new FileStorePathFactory.SstPathFactoryCache(pathFactory);
+        for (ManifestEntry entry : scan.withSnapshot(nextSnapshot.id()).plan().files()) {
```

Comment:
I can see large number of repeat reads. We can add a TODO to add manifest cache in scan.

Reply:
We can optimize this by expiring all snapshots at once so no repeated reads.